src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -388,6 +407,31 @@ private JsonRpcRequest ReadRequest(JToken json)
                 args is JArray argsArray ? (object)PartiallyParsePositionalArguments(argsArray) :
                 null;
 
+            // If method is $/progress, get the progress instance from the dictionary and call Report
+            string method = json.Value<string>("method");
+
+            if (method.Equals("$/progress"))
+            {
+                JToken progressId =
+                    args is JObject ? args["token"] :
+                    args is JArray ? args[0] :
+                    null;
+
+                JToken value =
+                    args is JObject ? args["value"] :
+                    args is JArray ? args[1] :
+                    null;
+
+                if (this.progressMap.TryGetValue(progressId.Value<long>(), out object progress))
+                {
+                    Type iprogressOfTType = MessageFormatterHelper.FindIProgressOfT(progress.GetType());
+                    Type valueType = iprogressOfTType.GenericTypeArguments[0];
+                    var reportMethod = iprogressOfTType.GetRuntimeMethod(nameof(IProgress<int>.Report), new Type[] { valueType });
+                    object typedValue = value.ToObject(valueType);
+                    reportMethod.Invoke(progress, new object[] { typedValue });
+                }
```

Comment:
for later: this could be vastly improved as far as CPU and GC pressure cost. We can talk about options here once we're sure the design is sound.

Reply:
We want to avoid all the reflection we can in this likely frequented code. The only code that should remain in this `if` block would be the last two lines. The rest should be computed and cached in the `progressMap` when it was originally stored there. 

---
In reply to: [302243024](https://github.com/milopezc/vs-streamjsonrpc/pull/1#discussion_r302243024) [](ancestors = 302243024)